File: PhaseFieldNet/MatlabCode/CH2D.m
clc;
clear;
close all

%% Parameter Initialization

% Spatial Parameters
Nx=64; 
Ny=64; 
Lx=1; 
Ly=1; 
hx=Lx/Nx; 
hy=Ly/Ny;
x=linspace(-0.5*Lx+hx,0.5*Lx,Nx);
y=linspace(-0.5*Ly+hy,0.5*Ly,Ny);

% Interfacial energy constant
%m = 6; 
%eps1 = hx*m/(2*sqrt (2)*atanh(0.9)) ; 
epsilon=0.0125; 
Cahn=epsilon^2;

% Discrete Fourier Transform
p=2*pi/Lx*[0:Nx/2 -Nx/2+1:-1];
q=2*pi/Ly*[0:Ny/2 -Ny/2+1:-1];
p2=p.^2; 
q2=q.^2; 
[pp2,qq2]=ndgrid(p2,q2);

% Time Discritization
dt=0.0025; 
T=0.5; 
Nt=round(T/dt);
Np=200;
ns=Nt/Np;

%% Initial Condition
%u = -0.045 + 0.05*(2*rand(Nx,Ny)-1) ;
%u=0.05*(2*rand(Nx,Ny)-1);
tau = 2000;
alpha = 1;
u=GRF(alpha, tau, Nx);
%% Initial Preview
figure(1);
clf;
contourf(x,y,real(u'),[0 0]); 
axis image
axis([x(1) x(Nx) y(1) y(Ny)])
pause(2)

%% Update
for iter = 1:Nt
    disp(['iteration = ' num2str(iter)])
    disp(['Maximum = ' num2str(max(u, [],'all'))])
    disp(['Minimum = ' num2str(min(u, [],'all'))])
    u=real(u);
    s_hat=fft2(u)-dt*(pp2+qq2).*fft2(u.^3-3*u);
    v_hat=s_hat./(1.0+dt*(2.0*(pp2+qq2)+Cahn*(pp2+qq2).^2));
    u=ifft2(v_hat);
    if (mod(iter,ns)==0)
        contourf(x,y,real(u'),[0 0]);
        title(['\Delta t = '  num2str(iter)])
        axis image
        axis([x(1) x(Nx) y(1) y(Ny)])
        pause(0.1)
    end
end

File: PhaseFieldNet/MatlabCode/CH2D.py
import time
import numpy as np
import torch
import matplotlib.pyplot as plt
from GRF import GRF  # Assuming GRF function is provided

# Parameter Initialization

# Spatial Parameters
Nx, Ny = 64, 64
Lx, Ly = 1, 1
hx, hy = Lx / Nx, Ly / Ny

x = torch.linspace(-0.5 * Lx + hx, 0.5 * Lx, Nx)
y = torch.linspace(-0.5 * Ly + hy, 0.5 * Ly, Ny)
xx, yy = torch.meshgrid(x, y)

# Interfacial energy constant
epsilon = 0.0125
Cahn = epsilon ** 2

# Discrete Fourier Transform
p = 2 * np.pi / Lx * torch.cat([torch.arange(0, Nx // 2 + 1), torch.arange(-Nx // 2 + 1, 0)])
q = 2 * np.pi / Ly * torch.cat([torch.arange(0, Ny // 2 + 1), torch.arange(-Ny // 2 + 1, 0)])
p2 = p ** 2
q2 = q ** 2
pp2, qq2 = torch.meshgrid(p2, q2)

# Time Discretization
dt = 0.0025
T = 0.5
Nt = round(T / dt)
Np = 200
ns = Nt // Np

# Initial Condition
tau = 2000
alpha = 1
u = GRF(alpha, tau, Nx)  # Initial condition using the GRF function
u = torch.tensor(u, dtype=torch.float32)

# Move to GPU if available
device = torch.device('cuda:2' if torch.cuda.is_available() else 'cpu')
u = u.to(device)
pp2, qq2 = pp2.to(device), qq2.to(device)

# Time Evolution
t_start = time.time()
for iteration in range(Nt):
    u = torch.real(u)
    s_hat = torch.fft.fft2(u) - dt * (pp2 + qq2) * torch.fft.fft2(u ** 3 - 3 * u)
    v_hat = s_hat / (1.0 + dt * (2.0 * (pp2 + qq2) + Cahn * (pp2 + qq2) ** 2))
    u = torch.fft.ifft2(v_hat)

Dt = time.time() - t_start
print("Elapsed time =", Dt)

